File: leetcode/156. BinaryTree Upside Down.go
type Stack struct {
	nodes []*TreeNode
}

// Push adds an element to the stack
func (s *Stack) Push(node *TreeNode) {
	s.nodes = append(s.nodes, node)
}

func (s *Stack) HasItems() bool {
	return len(s.nodes) > 0
}

// Pop removes and returns the top element of the stack
func (s *Stack) Pop() *TreeNode {
	if len(s.nodes) == 0 {
		return nil
	}
	node := s.nodes[len(s.nodes)-1]
	s.nodes = s.nodes[:len(s.nodes)-1]
	return node
}

// upsideDownBinaryTree function
func upsideDownBinaryTree(root *TreeNode) *TreeNode {
	if root == nil || root.Left == nil {
		return root
	}

	stack := &Stack{}
	current := root
	for current != nil {
		stack.Push(current)
		current = current.Left
	}

	current = stack.Pop()
	newRoot := current

	for stack.HasItems() {
		parent := stack.Pop()
		current.Left = parent.Right
		current.Right = parent
		parent.Left = nil
		parent.Right = nil
		current = parent
	}

	return newRoot
}

File: leetcode/156. BinaryTree Upside Down.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var upsideDownBinaryTree = function (root) {
    if (!root || !root.left) {
        return root;
    }
    const stack = [];
    let current = root;
    while (current) {
        stack.push(current);
        current = current.left;
    }
    current = stack.pop();
    const newRoot = current;
    while (stack.length > 0) {
        let parent = stack.pop();
        current.left = parent.right;
        current.right = parent;
        parent.left = null;
        parent.right = null;
        current = parent;
    }
    return newRoot;
};

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var upsideDownBinaryTree = function (root) {
    if (!root || !root.left) {
        return root;
    }
    const newRoot = upsideDownBinaryTree(root.left);
    root.left.left = root.right;
    root.left.right = root;
    root.left = null;
    root.right = null;
    return newRoot;
};
